app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -64,118 +60,92 @@ class TorManager @Inject constructor(
-
         Intent(app, ShareService::class.java).also { intent ->
             startForegroundService(app, intent)
         }
 
         tor.start()
+        changeStartingState(5)
+        if (settingsManager.automaticBridges.value) {
+            startCheckJob = launch {
+                LOG.info("Starting check job")
+                checkStartupProgress()
+                LOG.info("Check job finished")
+            }
+        } else {
+            val customBridges = settingsManager.customBridges.value
+            if (customBridges.isNotEmpty()) {
+                LOG.info("Using ${customBridges.size} custom bridges...")
+                tor.enableBridges(customBridges.map { "Bridge $it" })
+            }
+        }
+        tor.enableNetwork(true)
     }
 
-    fun stop(failedToConnect: Boolean = false) {
+    fun stop() {
         LOG.info("Stopping...")
         startCheckJob?.cancel()
         startCheckJob = null
-        _state.value = TorState.Stopping(failedToConnect)
         tor.stop()
         Intent(app, ShareService::class.java).also { intent ->
             app.stopService(intent)
         }
     }
 
-    @OptIn(DelicateCoroutinesApi::class)
     override fun onState(s: TorWrapper.TorState) {
         when (s) {
             NOT_STARTED -> LOG.info("new state: not started")
             STARTING -> LOG.info("new state: starting")
-            STARTED -> GlobalScope.launch {
-                LOG.info("new state: started")
-                try {
-                    onTorServiceStarted()
-                } catch (e: Exception) {
-                    LOG.warn("Error while starting Tor: ", e)
-                    stop()
-                }
-            }
-
+            STARTED -> LOG.info("new state: started")
             CONNECTING -> LOG.info("new state: connecting")
             CONNECTED -> LOG.info("new state: connected")
             DISABLED -> LOG.info("new state: network disabled")
             STOPPING -> LOG.info("new state: stopping")
-            STOPPED -> onStopped()
+            STOPPED -> {
+                LOG.info("new state: stopped")
+                _state.value = TorState.Stopped
+            }
         }
     }
 
-    @UiThread
-    fun onStopped() {
-        LOG.info("Stopped")
-        val failedToConnect = (state.value as? TorState.Stopping)?.failedToConnect == true
-        _state.value = TorState.Stopped(failedToConnect)
+    fun publishOnionService(port: Int) {
+        LOG.info("Starting hidden service...")
+        tor.publishHiddenService(port, 80, null)
     }
 
     override fun onBootstrapPercentage(percentage: Int) {
-        changeStartingState((percentage * 0.7).roundToInt())
+        changeStartingState(5 + (percentage * 0.9).roundToInt())
     }
 
     override fun onHsDescriptorUpload(onion: String) {
-        if (state.value !is TorState.Started) changeStartingState(90, onion)
-        onDescriptorUploaded(onion)
+        if (state.value !is TorState.Published) {
```

Comment:
Possible race with changing the state to Stopping and Stopped at shutdown?

Reply:
Good point. We are now only setting the `Stopped` state in `onState()`. So throwing the question back: Would Onionwrapper call `onHsDescriptorUpload()` at the same time?

Reply:
Yes, generally I think it's possible for onionwrapper to call back into the observer while the app's calling `TorWrapper` methods. In some cases I guess this is expected - for example we expect to observe the state changing to `STOPPING` and then `STOPPED` while `stop()` is being called. But possible to observe a descriptor upload too, I think.

Reply:
So do you suggest wrapping the `MutableStateFlow` in an `AtomicReference` or using a second atomic reference?

Reply:
I don't know what the semantics of wrapping a `MutableStateFlow` in an `AtomicReference` (or vice versa) would be. So I'd suggest something like this:

* Use an `AtomicReference<TorState>` to keep track of the state internally. This allows us to use `compareAndSet()` to avoid unwanted state changes (for example, ignore the descriptor upload unless the current state is `Starting`)
* Use a `MutableStateFlow<TorState>` to publish the state for external consumption. Whenever we update the `AtomicReference<TorState>`, update the `MutableStateFlow<TorState>` to let any observers see the change

This seems kind of fiddly and it's tempting to suggest wrapping the `AtomicReference` and the `MutableStateFlow` in a helper class that ensures we always update the `MutableStateFlow` when we update the `AtomicReference`. But YAGNI sayeth no.

Alternatively, maybe `MutableStateFlow` already has compare-and-set semantics that we can use?

Reply:
> Alternatively, maybe MutableStateFlow already has compare-and-set semantics that we can use?

It does.